core/src/main/java/com/digitalpebble/stormcrawler/protocol/okhttp/HttpProtocol.java

```diff
@@ -270,80 +285,105 @@ public ProtocolResponse getProtocolOutput(String url, final Metadata metadata) t
         if (completionTimeout != -1) {
             endDueFor = System.currentTimeMillis() + (completionTimeout * 1000);
         }
-        final ByteArrayBuffer buffer = new ByteArrayBuffer(bufferInitSize);
-        final byte[] tmp = new byte[4096];
-        int lengthRead;
-        while ((lengthRead = instream.read(tmp)) != -1) {
-            // check whether we need to trim
-            if (maxContent != -1 && buffer.length() + lengthRead > maxContent) {
-                buffer.append(tmp, 0, maxContent - buffer.length());
-                trimmed.setValue(true);
+
+        BufferedSource source = responseBody.source();
+        int bytesRequested = 0;
+        int bufferGrowStepBytes = 8192;
+
+        while (source.buffer().size() < maxContentBytes) {
+            bytesRequested += Math.min(bufferGrowStepBytes,
+                    (maxContentBytes - bytesRequested));
+            boolean success = false;
+            try {
+                success = source.request(bytesRequested);
+            } catch (IOException e) {
+                // requesting more content failed, e.g. by a socket timeout
+                if (partialContentAsTrimmed && source.buffer().size() > 0) {
+                    // treat already fetched content as trimmed
+                    trimmed.setValue(TrimmedContentReason.DISCONNECT);
+                    LOG.debug("Exception while fetching {}", e);
+                } else {
+                    throw e;
+                }
+            }
+            if (!success) {
+                // source exhausted, no more data to read
                 break;
             }
-            buffer.append(tmp, 0, lengthRead);
-            // check whether we hit the completion timeout
+
             if (endDueFor != -1 && endDueFor <= System.currentTimeMillis()) {
-                trimmed.setValue(true);
+                // check whether we hit the completion timeout
+                trimmed.setValue(TrimmedContentReason.TIME);
                 break;
             }
+
+            // okhttp may fetch more content than requested, quickly "increment"
+            // bytes
+            bytesRequested = (int) source.buffer().size();
         }
-        return buffer.toByteArray();
+        int bytesBuffered = (int) source.buffer().size();
+        int bytesToCopy = bytesBuffered;
+        if (maxContent != -1 && bytesToCopy > maxContent) {
+            // okhttp's internal buffer is larger than maxContent
+            trimmed.setValue(TrimmedContentReason.LENGTH);
+            bytesToCopy = maxContentBytes;
+        }
+        byte[] arr = new byte[bytesToCopy];
+        source.buffer().readFully(arr);
+        return arr;
     }
 
     class HTTPHeadersInterceptor implements Interceptor {
 
+        final ZoneId TIME_ZONE_UTC = ZoneId.of(ZoneOffset.UTC.toString());
+
         @Override
         public Response intercept(Interceptor.Chain chain) throws IOException {
+
+            String startFetchTime = DateTimeFormatter.ISO_INSTANT
```

Comment:
isn't that the same as Instant.now().toString()? the formatter used is
https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html#ISO_INSTANT

Reply:
Looks like, although I don't like the toString() methods which often implicitly assumes some localization. It's also not the best design to do the date formatting defined by the WARC format in the protocol layer. But re-parsing the date from a String is also not ideal.

Reply:
the protocol could store the value as a long and let the WARC code format it whichever way it wants e.g. using _Instant.ofEpochSecond(long epochSecond)_